test/langtools/tools/javac/annotations/typeAnnotations/TypeAnnosOnConstructorsTest.java

```diff
@@ -0,0 +1,145 @@
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8320001
+ * @summary compiler crash while adding type annotations to constructor's return type
+ * @library /tools/lib /tools/javac/lib
+ * @modules
+ *      jdk.compiler/com.sun.tools.javac.api
+ *      jdk.compiler/com.sun.tools.javac.main
+ * @build toolbox.ToolBox toolbox.JavacTask
+ * @run main TypeAnnosOnConstructorsTest
+ */
+
+import java.io.IOException;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import java.util.List;
+import java.util.Set;
+import javax.annotation.processing.AbstractProcessor;
+import javax.annotation.processing.RoundEnvironment;
+import javax.annotation.processing.SupportedAnnotationTypes;
+import javax.lang.model.SourceVersion;
+import javax.lang.model.element.TypeElement;
+
+import toolbox.JavacTask;
+import toolbox.Task;
+import toolbox.Task.Mode;
+import toolbox.Task.OutputKind;
+import toolbox.TestRunner;
+import toolbox.ToolBox;
+
+public class TypeAnnosOnConstructorsTest extends TestRunner {
+    protected ToolBox tb;
+
+    TypeAnnosOnConstructorsTest() {
+        super(System.err);
+        tb = new ToolBox();
+    }
+
+    public static void main(String... args) throws Exception {
+        new TypeAnnosOnConstructorsTest().runTests();
+    }
+
+    protected void runTests() throws Exception {
+        runTests(m -> new Object[]{Paths.get(m.getName())});
+    }
+
+    Path[] findJavaFiles(Path... paths) throws IOException {
+        return tb.findJavaFiles(paths);
+    }
+
+    @Test
+    public void testAnnoOnConstructors(Path base) throws Exception {
+        Path src = base.resolve("src");
+        Path y = src.resolve("Y.java");
+        Path yWrapper = src.resolve("YWrapper.java");
+
+        Path classes = base.resolve("classes");
+
+        Files.createDirectories(classes);
+
+        tb.writeJavaFiles(src,
+                """
+                import java.lang.annotation.Target;
+                import java.lang.annotation.ElementType;
+                import java.lang.annotation.Retention;
+                import java.lang.annotation.RetentionPolicy;
+
+                class Y {
+                    @TA public Y() {}
+                }
+
+                @Target(ElementType.TYPE_USE)
+                @Retention(RetentionPolicy.RUNTIME)
+                @interface TA {}
+                """,
+                """
+                class YWrapper {
+                    // just to make the compiler load the class file for class Y above
+                    Y y;
+                }
+                """);
+
+        // we need to compile Y first
+        new JavacTask(tb)
+                .files(y)
+                .outdir(classes)
+                .run();
+
+        Path classDir = getClassDir();
+        new JavacTask(tb)
+                .classpath(classes, classDir)
+                .options("-processor", SimpleProcessor.class.getName())
+                .files(yWrapper)
```

Comment:
It should be equivalent to use `.classes("Y")` to request annotation processing of `Y` directly, and avoid the need for `YWrapper`. But either way is fine.

Reply:
thanks for your comment, yes I tried that but all of the APIs we offer in JavacTasks were failing stating that the file was not a source file, this is why I decided to use a wrapper

Reply:
yup your are right, will fix the test, it seems like I tried another combination of options